### ISSUE TITLE
Hold for #38909 [$500] IOU - Map is blinking when navigate to Distance request

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#38838]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.56-0
**Reproducible in staging?:** y
**Reproducible in production?:** n
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to FAB> +> Distance> Complete the IOU flow
3. Navigate to created Distance request
4. Go to details page and go back to Distance card page
5. Repeat the step 4 and take a not of map behavior

## Expected Result:
Map should be stable

## Actual Result:
Map is blinking when navigate to created Distance request

## Workaround:
n/a

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/5f4e7922-6b5e-4a5d-ad40-fa5a3af1317f

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b572848dc86f02bf</li>
        <li>Upwork Job ID: 1771279748516020224</li>
        <li>Last Price Increase: 2024-03-29</li>
<li>Automatic offers: </li>
<ul>
<li>paultsimura | Contributor | 0</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
We think that this bug might be related to #wave-collect RELEASE 1

**Comment by jasperhuangg** [comment]
Definitely don't need to block deploy as this is a display bug that doesn't really break usage of this feature

**Comment by lanitochka17** [comment]
During testing, another issue was found:
Map reloads with "Route pending" when clicking Save button without any changes

https://github.com/Expensify/App/assets/78819774/707e43cd-1204-4dc3-9ff4-a49e2fc449d4

Is this the same issue or do I need to open a separate one?

**Comment by ebcabaybay** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Map is blinking when navigating to created Distance request

### What is the root cause of that problem?
In ReportScreen.tsx, this useEffect() triggers unneeded re-render to the map image component when Report.openReport(report.reportID) is called.

```
useEffect(() => {
        if (!isFocused || prevIsFocused || !ReportUtils.isChatThread(report) || report.notificationPreference !== CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN) {
            return;
        }
        Report.openReport(report.reportID);

        // We don't want to run this useEffect every time `report` is changed
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [prevIsFocused, report.notificationPreference, isFocused]);
```


### What changes do you think we should make in order to solve the problem?
Add a new condition called isMoneyRequestView that determines if the current screen is a MoneyRequestView

Add this new condition to this line to immediately return instead of calling Report.openReport   

```
if (isMoneyRequestView || !isFocused || prevIsFocused || !ReportUtils.isChatThread(report) || report.notificationPreference !== CONST.REPORT.NOTIFICATION_PREFERENCE.HIDDEN) {
            return;
        }
```

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Map is blinking when navigate to Distance request.

### What is the root cause of that problem?

**TLDR:** The RCA was already mentioned in a previous [proposal](https://github.com/Expensify/App/issues/38838#issuecomment-2016581354).

Offending PR https://github.com/Expensify/App/pull/30269 introduced the following logic:

https://github.com/Expensify/App/blob/fed9d6c0f4cb47605233c556ea1794bec002f6f8/src/pages/home/report/ReportActionsView.tsx#L135-L141

which generates a new `listID` everytime `OpenReport` is called (`isLoadingInitialReportActions`).

The problem lies in the fact that `listID` is passed as `key` to the FlatList component [here](https://github.com/Expensify/App/blob/fed9d6c0f4cb47605233c556ea1794bec002f6f8/src/pages/home/report/ReportActionsList.tsx#L622), which force re-renders the entire component (with all its children), creating the flicker behaviour.

### What changes do you think we should make in order to solve the problem?

We should pass the `listID` to the FlatList's [`extraData`](https://reactnative.dev/docs/flatlist#extradata) array [here](https://github.com/Expensify/App/blob/fed9d6c0f4cb47605233c556ea1794bec002f6f8/src/pages/home/report/ReportActionsList.tsx#L532) instead of passing it as a `Reeact.key` to the FlatList.

<details>
<summary>Why use extraData instead of key ?</summary>

.
`extraData` should be used in a `FlatList` to inform the component to re-render when there is a change in state or props outside of the `data` prop, and the appearance of our items depends on this state or props. Since `extraData` is explicitly designed for this purpose, when it changes, React Native will only re-render the items that are affected by the change. This is generally efficient and avoids unnecessary renders.

On the other hand, the `key` prop is meant for a unique identifier for the entire list to help React with the reconciliation process. It is not intended to be used as a mechanism to control the re-render of a `FlatList`. Changing keys as a way to force a re-render is generally not recommended because it tells React that the entire component (with all its children) is completely different from the one before. This causes React to unmount the old component and mount a new component instead of efficiently updating the existing one. This process is more expensive in terms of performance, can lead to a flicker as mentioned, and can reset internal component state.

Using the `key` prop on the `FlatList` itself to force rerenders can lead to a visible flicker because React will remount the entire list and its items, which is a more costly operation than simply re-rendering items that have changed. The remounting process would cause a complete rebuild of the list structure, leading to visible flicker as old items are removed and new ones are added.

</details>

### Videos (before / after)

<details>
<summary>MacOS: Chrome / Safari</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/c07f57e9-e4f5-4d9f-ac23-82f5cc27d521"> | <video src="https://github.com/Expensify/App/assets/56457735/a8190347-b5ca-4fad-b879-b207b8f365bc"> |

</details>

**Comment by ikevin127** [comment]
I'd like to ask @perunt as author of PR https://github.com/Expensify/App/pull/30269 why the `listID` was passed to FlatList `key` instead of being passed to the existing `extraData` FlatList prop.

It would be good to get some context on this so we can avoid a conflict between the PR's intended implementation and this issue's possible fix.

**Comment by ikevin127** [comment]
I think this should be put on hold for until the PR https://github.com/Expensify/App/pull/39040 is merged then re-test as I'm not entirely sure passing `listID` as (React) `key` to the FlatList is the best idea when we have the `extraData` prop. Also the mentioned PR is fixing https://github.com/Expensify/App/issues/38909 (issue), but might not fix this issue.

I commented on the PR [#39040 (comment)](https://github.com/Expensify/App/pull/39040/files#r1541354233) kindly asking for a comment to be added above the declaration of `listID` explaining its role and the reason behind why it's passed to the FlatList as `key`.

**Comment by perunt** [comment]
> I'd like to ask @perunt as author of PR https://github.com/Expensify/App/pull/30269 why the listID was passed to FlatList key instead of being passed to the existing extraData FlatList prop.

I'll dupe my message here as well:

So, here's the deal: I need to make the whole list refresh itself from scratch. It's all about the virtualization thing the list does. Basically, this refresh trick is mostly needed when you're linking comments, especially when jumping from looking at a bunch of messages to just a few. The virtualization in FlatList gets a bit wonky when the amount of stuff shown shrinks as you move to a specific message. So, to keep everything looking right, we gotta refresh the whole list, not just poke the FlatList to update what it's already showing. So if we put it to `extraData` the comment linking would lead to wrong place in some specific cases

**Comment by perunt** [comment]
Hey guys! Just a quick heads-up, this [PR](https://github.com/Expensify/App/pull/39040) has been merged. So this shouldn't be an issue anymore
